File: cpp/src/Decrypter.cpp
#include <Decrypter.h>

std::string Decrypter::decrypt(const std::vector<Uint> &enc)
{
  Uint k, c;
  std::string msg = "";

  for (auto &&l : enc)
  {
    k = static_cast<Uint>(1);
    c = static_cast<Uint>(l);
    for (Uint j = 0; j < this->d; ++j)
      k = (k * c) % this->n;
    msg += static_cast<char>(k);
  }

  return msg;
}

File: CryptoSocket/Crypto/Decrypter.py
#!/usr/bin/env python

import numpy as np

class Decrypter:

  def __init__(self, n, d):
    self.n = n
    self.d = d

  def decrypt(self, enc):
    msg = ''
    for l in enc:
      k = 1
      c = np.uint(l)
      for _ in range(self.d):
        k = np.uint((k * c) % self.n)
      msg += chr(k)
    return msg

  def __repr__(self):
    fmt_str = "Encoder object\n"
    members = [attr for attr in dir(self) if not callable(getattr(self, attr)) and not attr.startswith("__")]
    for m in members:
      fmt_str += '    %s: %s\n'%(m, eval('self.%s'%m))
    return fmt_str
